netrd/distance/distributional_nbd.py

```diff
@@ -0,0 +1,247 @@
+"""
+distributional_nbd.py
+------
+
+Distributional Non-backtracking Spectral Distance.
+
+"""
+
+import numpy as np
+import networkx as nx
+import scipy.sparse as sp
+from scipy.spatial.distance import euclidean, chebyshev
+
+from .base import BaseDistance
+
+
+class DistributionalNBD(BaseDistance):
+    """
+    Distributional Non-backtracking Spectral Distance.
+
+    Computes the distance between two graphs using the empirical spectral density
+    of the non-backtracking operator.
+
+    See: 
+    "Graph Comparison via the Non-backtracking Spectrum" 
+    A. Mellor & A. Grusovin
+    arXiv:1812.05457 / 10.1103/PhysRevE.99.052309
+
+    """
+
+    VECTOR_DISTANCES = {'euclidean': euclidean, 'chebyshev': chebyshev}
+
+    def dist(
+        self,
+        G1,
+        G2,
+        sparse=False,
+        shave=True,
+        keep_evals=False,
+        k=None,
+        vector_distance='euclidean',
+        **kwargs
+    ):
+        """
+        Distributional Non-backtracking Spectral Distance.
+
+        Parameters
+        ----------
+
+        G1, G2 (nx.Graph)
+            The two graphs to compare.
+
+        sparse (bool)
+            If sparse, matrices and eigenvalues found using sparse methods.
+            If sparse, parameter 'k' should also be specified.
+            Default: False
+
+        k (int)
+            The number of largest eigenvalues to be calculated for the
+            spectral density.
+
+        vector_distance (str)
+            The distance measure used to compare two empirical distributions.
+            Currently available are 'euclidean' and 'chebyshev', implemented 
+            using SciPy.
+            Default: 'euclidean'
+
+        keep_evals (bool)
+            If True, stores the eigenvalues of the reduced non-backtracking
+            matrix in self.results['evals']
+            Default: False
+
+
+        Returns
+        -------
+        float
+            The distance between `G1` and `G2`
+
+        """
+        B1 = reduced_hashimoto(G1, shave=shave, sparse=sparse, **kwargs)
+        B2 = reduced_hashimoto(G2, shave=shave, sparse=sparse, **kwargs)
+
+        # Find spectrum
+        evals1 = nb_eigenvalues(B1, k=k)
+        evals2 = nb_eigenvalues(B2, k=k)
+
+        # Save spectrum
+        if keep_evals:
```

Comment:
We've been thinking through better ways of doing this, since storing lots of metadata by default is admittedly less than ideal, but for consistency with the other distances can you default `keep_evals` to `True`?

Reply:
Fixed via 273cf3d

I agree that storing metadata by default should be avoided.
Eigenvalues are only useful for debugging/other interests - the main purpose is purely distance calculation.

Reply:
Thanks. I've opened an issue about storing intermediate results - tl;dr is that we had been doing it because it's useful for our own experiments but we've been considered restructuring things a bit to make this less necessary.